### PR TITLE
Use Theano's IfElse instead of Switch for in_train_phase

### PR DESCRIPTION
**Pull request by gvtulder**
Using `IfElse` should allow Theano to [perform lazy evaluation](http://deeplearning.net/software/theano/tutorial/conditions.html) of the training and testing branches, so in theory it might be faster. `Switch` always computes both outputs.

### CODE REVIEW

**Comment by EderSantana** [comment]
cool! which benchmarks did you run?

**Comment by gvtulder** [comment]
Benchmarks are a bit difficult. In most cases this won't give any measurable improvement, because the difference between the training and testing branches is usually quite small. Theano will merge them and there aren't that many unnecessary operations left to be skipped.

There will be an improvement when the training and testing phases use different operations that cannot be merged. For example, cuDNN has separate batch normalization functions for training and inference, and you would want to run only the one that is actually used.

I did a small experiment with the `mnist_cnn.py` example, to which I added two batch normalization layers. This also requires the batch normalization patches for Theano and Keras I am working on ([Keras branch](https://github.com/gvtulder/keras/tree/f-theano-new-batchnorm-interface) and [Theano branch](https://github.com/gvtulder/Theano/tree/f-batchnorm-abstract)).

The current Keras code uses a switch, which means that Theano will run both the training and testing batch normalization functions from cuDNN. Timings:
```
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 76s - loss: 0.3945 - acc: 0.8785 - val_loss: 0.3039 - val_acc: 0.9070
Epoch 2/12
60000/60000 [==============================] - 74s - loss: 0.1396 - acc: 0.9592 - val_loss: 0.0604 - val_acc: 0.9810
Epoch 3/12
60000/60000 [==============================] - 74s - loss: 0.1018 - acc: 0.9702 - val_loss: 0.0452 - val_acc: 0.9846
```

With `ifelse`, only the training op will be run, giving a small speed-up (most of the time is spent in the convolution layers and elsewhere):
```
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 70s - loss: 0.3955 - acc: 0.8784 - val_loss: 0.3160 - val_acc: 0.9004
Epoch 2/12
60000/60000 [==============================] - 68s - loss: 0.1395 - acc: 0.9593 - val_loss: 0.0635 - val_acc: 0.9796
Epoch 3/12
60000/60000 [==============================] - 69s - loss: 0.1022 - acc: 0.9701 - val_loss: 0.0446 - val_acc: 0.9844
```

**Comment by EderSantana** [comment]
well, 4 seconds can add up quick. Let us get the @farizrahman4u in the conversation. What do you think?

**Comment by fchollet** [comment]
That looks reasonable to me.

**Comment by farizrahman4u** [comment]
@fchollet In that case, we should use IFElse for K.switch() too.

**Comment by DingKe** [comment]
How about change Theano's swtich implementation and and another API for tf-like select for element-wise slection operation #4519

**Comment by gvtulder** [comment]
@farizrahman4u, @DingKe: You could use ifelse instead of switch if the condition is a scalar, so you'd have to check that first. K.switch (at least on Theano) accepts elementwise conditions as well.

Be aware that the Theano documentation has a warning that [ifelse is not always faster than switch](http://deeplearning.net/software/theano/tutorial/conditions.html), which is why the scalar-switch isn't optimized automatically by Theano. I don't know if this is still true, though, since the ticket it refers to seems to be quite old.
